cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/types/FunctionPointerType.kt

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2021, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.graph.types
+
+import de.fraunhofer.aisec.cpg.frontends.Language
+import de.fraunhofer.aisec.cpg.frontends.LanguageFrontend
+import de.fraunhofer.aisec.cpg.graph.edge.PropertyEdge
+import de.fraunhofer.aisec.cpg.graph.edge.PropertyEdge.Companion.propertyEqualsList
+import de.fraunhofer.aisec.cpg.graph.edge.PropertyEdge.Companion.transformIntoOutgoingPropertyEdgeList
+import de.fraunhofer.aisec.cpg.graph.edge.PropertyEdge.Companion.unwrap
+import de.fraunhofer.aisec.cpg.graph.edge.PropertyEdgeDelegate
+import de.fraunhofer.aisec.cpg.graph.types.PointerType.PointerOrigin
+import java.util.*
+import org.neo4j.ogm.annotation.Relationship
+
+/**
+ * FunctionPointerType represents FunctionPointers in CPP containing a list of parameters and a
+ * return type.
+ */
+class FunctionPointerType : Type {
+    @Relationship(value = "PARAMETERS", direction = Relationship.Direction.OUTGOING)
+    var parametersPropertyEdge: MutableList<PropertyEdge<Type>> = mutableListOf()
```

Comment:
While we are at it: Do we really need property edges here? Personally, I would try to reduce the number of property edges to cases where they really make sense and not just introduce them for the sake of having them. Here, I do not see a use case for them.

Reply:
Here I have to opposit opinion so I am not sure what to do, as far as I know the save the order of the parameter s which is an information that would get lost in Neo4j as we don't have the list index there. Would it be acceptable to keep them?

I am afraid we cannot solve that with a converter, we tried that before and there was no way to get the list and add properties for the single elements.

Reply:
In general, we should look at all the usages to identify their usages and eliminate unnecessary ones and implement the respective use-case for the edges we keep in the same PR. Right now, we have them almost everywhere and never make use of the properties which is quite useless. So let's say this is a problem of "very near future us"

Reply:
Ah I forgot about that :( Ok in this case we need to keep it like that. Maybe I'll have another look if this is somehow possible without introducing property edges (for the use case where its just the index).